dask/dataframe/io/parquet/arrow.py

```diff
@@ -0,0 +1,452 @@
+import json
+
+import pandas as pd
+import pyarrow as pa
+import pyarrow.parquet as pq
+from ....bytes.core import get_pyarrow_filesystem
+from ....utils import natural_sort_key, getargspec
+from ..utils import _get_pyarrow_dtypes, _meta_from_dtypes
+from ...utils import clear_known_categories
+
+from .utils import _parse_pandas_metadata, _normalize_index_columns, Engine
+
+
+def _get_md_row_groups(pieces):
+    """ Read file-footer metadata from each individual piece.
+
+    Since this operation can be p[ainfully slow in some cases, abort
+    if any metadata or statistics are missing
+    """
+    row_groups = []
+    for piece in pieces:
+        for rg in range(piece.get_metadata().num_row_groups):
+            row_group = piece.get_metadata().row_group(rg)
+            for c in range(row_group.num_columns):
+                if not row_group.column(c).statistics:
+                    return []
+            row_groups.append(row_group)
+    return row_groups
+
+
+class ArrowEngine(Engine):
+    @staticmethod
+    def read_partition(fs, piece, columns, index, **kwargs):
+        """ Read a single piece of a Parquet dataset into a Pandas DataFrame
+
+        This function is called many times in individual tasks
+
+        Parameters
+        ----------
+        fs: FileSystem
+        piece: object
+            This is some token that is returned by Engine.read_metadata.
+            Typically it represents a row group in a Parquet dataset
+        columns: List[str]
+            List of column names to pull out of that row group
+        index: str, List[str], or False
+            The index name(s).
+        **kwargs:
+            Includes `"kwargs"` values stored within the `parts` output
+            of `pyarrow.read_metadata`: `partitions` and `categories`.
+
+        Returns
+        -------
+        A Pandas DataFrame
+        """
+        partitions = kwargs["partitions"]
+        categories = kwargs["categories"]
+        read_kwargs = kwargs.get("read", {})
```

Comment:
Should we include these directly in the signature above?

Reply:
Why `read_kwargs` rather than passing through `kwargs directly?

Reply:
I'm actually playing around with this right now, and I'm not certain of the cleanest way to get what we need - I want to allow users to provide arguments ito `to_parquet` and `read_parquet` that will be passed along to a *specific* underlying engine functions (e.g. only `ParquetDataset`, and not `piece.read`  for `pyarrow`).   I am thinking that one solution is to requre the user to pass in dict of dicts, to ensure that these pass-through params are included when the top-level key matches the name of the function that is called.   What is the conventional way to deal with this issue?

Reply:
So, for `partitions` and `categories` I think it's unambiguous that they should be handled in the signature itself.

I can see a concern if there is ambiguity about where things should go.  This is similar today to how we handle `storage_options=` in the Dask bytes backends for systems like S3/GCS/HDFS.  

For the particular cses of `"read"` and `"ParquetDataset"` I'm concerned that users won't know the internals of Arrow or FastParquet well enough to know these terms ahead of time.  I wonder if there are some more general terms that we might be able to use more generically and call them out specifically in the general docstring.

Reply:
@martindurant do you have thoughts here on how to bundle up some keywords consistently to pass along to particular components of the pipeline here?

Reply:
I'm not sure that there's any getting around users having to read up on the particular engine implementation when using features which are "expert" or otherwise not consistent across engines. I think if we document enough of the common ones, and show lots of examples, that's fine. This is similar with the case of csv or any other place we defer to an external library/ies, particularly the set of file-systems and all their credentials.

To the concrete question, we already name the components of the engine by the method names, so it would be reasonable to have several *_kwargs for each piece. That way, we don't need to keep track of any signature changes in the dependent packages.
However, I think the kwarg names will always be mutually exclusive between the parts of a given call, so the second possibility is to allow **kwargs in every method and `pop` off the one ones of interest to each function; the third is the same, except to include the expected kwargs in the method signatures (which is good information) and ignore any extras that might be passed through. I think I prefer the final model.